doc/src/devdocs/gc-sa.md

```diff
@@ -0,0 +1,282 @@
+### NOTSAFEPOINT
+
+This is perhaps the most common annotation, and should be placed on any function
+that is known not to possibly lead to reaching a GC safepoint. In general, is is
+only safe for such a function to perform arithmetic, memory accesses and calls to
+functions either annotated NOTSAFEPOINT or otherwise known not to be safepoints (e.g.
+function in the C standard library, which are hardcoded as such in the analyzer)
+
+It is valid to keep values unrooted across calls to any function annotated with this
+attribute:
+
+Usage Example:
+```c
+void jl_get_one() NOTSAFEPOINT {
+  return 1;
+}
+
+jl_value_t *example() {
+  jl_value_t *val = jl_alloc_whatever();
+  // This is valid, even though `val` is unrooted, because
+  // jl_get_one is not a safepoint
+  jl_get_one();
+  return val;
+}
+```
+
+### MAYBE_UNROOTED/ROOTS_TEMPORARILY
+
+When `MAYBE_UNROOTED` is annotated as an argument on a function,
+indicates that said argument may be passed, even if it is not rooted.
+In the ordinary course of events,
+the julia ABI guarantees that callers root values before passing them to
+callees. However, some functions do not follow this ABI and allow values
+to be passed to them even though they are not rooted. Note however, that
+this does not automatically imply that said argument will be preserved.
+The `ROOTS_TEMPORARILY` annotation provides the stronger guarantee that,
+not only may the value be unrooted when passed, it will also be preserved
+across any internal safepoints by the callee.
+
+Note that NOTSAFEPOINT essentially implies MAYBE_UNROOTED/ROOTS_TEMPORARILY,
+because the rootedness of an argument is irrelevant if the function contains
+no safepoints.
+
+One additional point to note is that these annotations apply on both the
+caller and the callee side. On the caller side, they lift rootedness
+restrictions that are normally required for julia ABI functions. On
+the callee side, they have the reverse effect of preventing these arguments
+from being considered implicitly rooted.
+
+If either of these annotations is applied to the function as a whole, it applies
+to all arguments of the function. This should generally only be necessary for
+varargs functions.
+
+Usage example:
+```
+JL_DLLEXPORT void JL_NORETURN jl_throw(jl_value_t *e MAYBE_UNROOTED);
+jl_value_t *jl_alloc_error();
+
+void example() {
+  // The return value of the allocation is unrooted. This would normally
+  // be an error, but is allowed because of the above annotation.
+  jl_throw(jl_alloc_error());
+}
+```
+
+### PROPAGATES_ROOT
+
+This annotation is commonly found on accessor functions that return one rootable
+object stored within another. When annotated on a function argument, it tells
+the analyzer that the root for that argument also applies to the value returned
+by the function.
+
+Usage Example:
+```c
+jl_value_t *jl_svecref(jl_svec_t *t PROPAGATES_ROOT, size_t i) NOTSAFEPOINT;
+
+size_t example(jl_svec_t *svec) {
+  jl_value_t *val = jl_svecref(svec, 1)
```

Comment:
This is only valid if `svec` can't be mutated. It's certainly OK for objects that does not escape and it is OK in this particular case only because `jl_gc_safepoint` is defined to not synchronize with any julia code. Any other safepoint here will be invalid. Does the analysis code handle these logic?

Reply:
No, that's a potential future refinement.